### PR TITLE
Add taget factory for splitting raster into tiles

### PR DESCRIPTION
**Pull request by Aariq**
I took a shot at creating a `tar_terra_tiles()` which is a target factory that creates three targets:
1) an upstream target that takes a SpatRaster and a template to define the tiles (the `y` argument of `terra::makeTiles()`), splits the raster into tiles, saves them to disk, and returns a vector of file paths.
2) a file target that maps over the file paths returned by `makeTiles()` using a target with format = "file" or "file_fast"
3) a downstream target that maps over the file paths and reads them in as rasters using dynamic branching.

This downstream target is a pattern that is ready to use in further targets with `pattern = map()` and starts to address #74.

This is heavily inspired by `tarchetypes::tar_files()`

The main downside to this pattern is that it basically duplicates the tiles on disk—once saved in a directory defined by the `tiles_dir` argument and again in the targets store when those files are read back in.  Deleting those files would invalidate the files target, but not the upstream target that creates them, so I'm forced to have the upstream target always re-run with `cue = tar_cue("always")`.  This is not ideal as making tiles is often a computationally expensive step.  I've left `#TODO` comments with questions and aspirations about how this might work better and would love feedback.

### CODE REVIEW

**Comment by Aariq** [comment]
Ok, it is "working" now (not passing check, but there is a test that should work now to give you an idea of what it does).  Here's what the manifest looks like:

``` r
library(targets)
tar_script({
    library(targets)
    library(geotargets)
    # devtools::load_all()
    library(terra)
    list(
        tar_target(
            my_file,
            system.file("ex/elev.tif", package="terra"),
            format = "file"
        ),
        tar_terra_rast(
            my_map,
            terra::rast(my_file)
        ),
        tar_terra_tiles(
            name = rast_split,
            raster = my_map,
            template = terra::rast(ncols = 2, nrows = 2),
            tiles_dir = "tiles_test"
        )
    )
})
targets::tar_manifest(callr_function = NULL)
#> terra 1.7.71
#> # A tibble: 5 × 3
#>   name             command                                               pattern
#>   <chr>            <chr>                                                 <chr>  
#> 1 my_file          "system.file(\"ex/elev.tif\", package = \"terra\")"   <NA>   
#> 2 my_map           "terra::rast(my_file)"                                <NA>   
#> 3 rast_split_tile  "geotargets:::make_tiles(raster = my_map, template =… <NA>   
#> 4 rast_split_files "rast_split_tile"                                     map(ra…
#> 5 rast_split       "rast(rast_split_files)"                              map(ra…
```

<sup>Created on 2024-05-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by Aariq** [comment]
Just pushed a fairly major improvement:
- No longer requires writing out to files outside of `_targets/`
- Now creates just two targets: 
    - an upstream target that is a list of numeric vectors that can be coerced to SpatExtent objects
    - a downstream target that iterates over the list of numeric vectors using them to set the `window()` of the original raster
- Exports two new helper functions, mostly so the command in `tar_manifest()` looks pretty, but also they might be helpful on their own.

File: R/tar_terra_tiles.R [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+#' @export
+#'
+#' @examples
+#' f <- system.file("ex/elev.tif", package="terra")
+#' r <- terra::rast(f)
+#' create_tile_exts(r, ncol = 2, nrow = 2)
+create_tile_exts <- function(raster, ncol, nrow) {
+    template <- terra::rast(terra::ext(raster), ncol = ncol, nrow = nrow, crs = terra::crs(raster))
+    tile_ext <- terra::getTileExtents(raster, template)
+    lapply(1:nrow(tile_ext), \(i) tile_ext[i,])
```
Comment (njtierney) [inline R/tar_terra_tiles.R:260]:
```suggestion
    lapply(seq_len(nrow(tile_ext)), \(i) tile_ext[i,])
```

File: tests/testthat/test-tar_terra_tiles.R
Comment (njtierney) [inline tests/testthat/test-tar_terra_tiles.R:1]: Small thing, but all the tests use hyphens to separate the words, can this change to hyphen separated?

File: R/tar_terra_tiles.R [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        storage = storage,
+        retrieval = retrieval,
+        cue = cue,
+        description = description
+    )
+
+    list(windows, tiles)
+}
+
+#' Copy a raster with a window
```
Comment (njtierney) [inline R/tar_terra_tiles.R:206]:
```suggestion
#' Copy a raster within a window
```

File: R/tar_terra_tiles.R [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+#'
+#' Create a new SpatRaster object based on an original SpatRaster and a window.
```
Comment (njtierney) [inline R/tar_terra_tiles.R:208]:
```suggestion
#' Create a new SpatRaster object as specified by a window (area of interest) over the original SpatRaster.
```

File: R/tar_terra_tiles.R [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+#' This is a wrapper around [terra::window()] which, rather than modifying the
+#' SpatRaster in place, returns a new SpatRaster leaving the original unchanged.
+#'
+#' @param raster A SpatRaster object
+#' @param window A SpatExtent object
```
Comment (njtierney) [inline R/tar_terra_tiles.R:213]:
```suggestion
#' @param window A SpatExtent object. The window of interest.
```

**Review by njtierney — CHANGES_REQUESTED** [review]
Just some small changes, this is looking really nice!

File: R/tar_terra_tiles.R [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+#'
+#' @param raster a `SpatRaster` object to be split into tiles
+#' @param ncol ncol
+#' @param nrow nrow
+#' @param filetype character. File format expressed as GDAL driver names passed
+#'   to [terra::makeTiles()]
+#' @param gdal character. GDAL driver specific datasource creation options
+#'   passed to [terra::makeTiles()]
+#'
+#' @param ... additional arguments not yet used
```
[the remark this answers is not in the document]
Comment (Aariq) [reply R/tar_terra_tiles.R:15]: This was for consistency with other functions `...` arg and I think it is probably grammatically correct as is, although could be "additional arguments which are not yet used".  Either way, I'll keep it as-is for now and if we want to change it we can change it for all functions in a separate PR.

[the remark this answers is not in the document]
Comment (njtierney) [reply R/tar_terra_tiles.R:15]: Yup that makes sense!

**Review by njtierney — APPROVED** [review]
Looks good to me!

File: R/tar_terra_tiles.R (hunk 4 above)
Comment (njtierney) [reply R/tar_terra_tiles.R]: Yup that reads great!

**Pull request merged**